### PR TITLE
Resolve catching exceptions TODOs

### PR DESCRIPTION
**Pull request by shreyamalviya**
# What does this PR do?

Fixes a part of #1826 

## PR Checklist
* [x] Have you added an explanation of what your changes do and why you'd like to include them?
* [x] Is the TravisCI build passing?
* [ ] ~Was the CHANGELOG.md updated to reflect the changes?~
* [ ] ~Was the documentation framework updated to reflect the changes?~

## Testing Checklist

* [x] Added relevant unit tests?
* [ ] ~Have you successfully tested your changes locally? Elaborate:~
    > ~Tested by {Running the Monkey locally with relevant config/running Island/...}~
* [ ] ~If applicable, add screenshots or log transcripts of the feature working~

### CODE REVIEW

File: monkey/infection_monkey/master/ip_scanner.py [hunk 1]
```diff
@@ -61,16 +61,30 @@ def _scan_addresses(
...
+                    ping_scan_data = self._puppet.ping(address.ip, icmp_timeout)
+                except Exception as ex:
+                    logger.debug(f"Exception encountered when pinging {address.ip}: {str(ex)}")
+
+                try:
+                    port_scan_data = self._puppet.scan_tcp_ports(address.ip, tcp_ports, tcp_timeout)
+                except Exception as ex:
+                    logger.debug(
+                        f"Exception encountered when scanning TCP ports on {address.ip}: {str(ex)}"
+                    )
```
Comment (mssalvatore) [inline monkey/infection_monkey/master/ip_scanner.py:74]: If an exception is thrown, `ping_scan_data` or `port_scan_data` would be `None`. It's probably safer to return data structures with default values.

We can add a unit test for this case.

**Review by mssalvatore — CHANGES_REQUESTED** [review]

File: monkey/infection_monkey/master/automated_master.py [hunk 2]
```diff
@@ -225,8 +225,10 @@ def _run_plugins(
 
         interrupted_message = f"Received a stop signal, skipping remaining {plugin_type}s"
         for p in interruptible_iter(plugins, self._stop, interrupted_message):
-            # TODO: Catch exceptions to prevent thread from crashing
-            callback(p)
+            try:
+                callback(p)
+            except Exception as ex:
+                logger.debug(f"Exception encountered when running {plugin_type} {p}: {str(ex)}")
```
Comment (VakarisZ) [inline monkey/infection_monkey/master/automated_master.py:231]: Why do we want to catch all exceptions here? And most of all why we log uncaught exceptions as debug level?

File: monkey/infection_monkey/master/ip_scanner.py [hunk 3]
```diff
@@ -61,16 +61,30 @@ def _scan_addresses(
                 address = addresses.get_nowait()
                 logger.info(f"Scanning {address.ip}")
 
-                # TODO: Catch exceptions to prevent thread from crashing
-                ping_scan_data = self._puppet.ping(address.ip, icmp_timeout)
-                port_scan_data = self._puppet.scan_tcp_ports(address.ip, tcp_ports, tcp_timeout)
+                try:
+                    ping_scan_data = self._puppet.ping(address.ip, icmp_timeout)
+                except Exception as ex:
```
Comment (VakarisZ) [inline monkey/infection_monkey/master/ip_scanner.py:66]: Same questions. Why do we catch all exceptions? Why we log it on debug level?

**Review by VakarisZ — CHANGES_REQUESTED** [review]
I don't understand what's the point of catching all exceptions and logging them in debug level

File: monkey/infection_monkey/master/automated_master.py (hunk 2 above)
Comment (shreyamalviya) [reply monkey/infection_monkey/master/automated_master.py:231]: So the thread doesn't crash? Will change the log level to warning.

File: monkey/infection_monkey/master/ip_scanner.py [hunk 4]
```diff
@@ -61,16 +61,33 @@ def _scan_addresses(
...
+                    logger.warning(f"Exception encountered when pinging {address.ip}: {str(ex)}")
```
Comment (VakarisZ) [inline monkey/infection_monkey/master/ip_scanner.py:67]: It's not a warning it's an error. Ping scanner should handle it's errors

File: monkey/infection_monkey/master/ip_scanner.py [hunk 5]
```diff
@@ -61,16 +61,33 @@ def _scan_addresses(
...
+                    ping_scan_data = PingScanData(False, None)
+
+                try:
+                    port_scan_data = self._puppet.scan_tcp_ports(address.ip, tcp_ports, tcp_timeout)
+                except Exception as ex:
```
Comment (VakarisZ) [inline monkey/infection_monkey/master/ip_scanner.py:72]: @mssalvatore why are we doing this?

Comment (mssalvatore) [reply monkey/infection_monkey/master/ip_scanner.py:72]: If these raise exceptions, the thread will crash and become unusable. If we can ensure that `network_scanning.scan_tcp_ports()`, `network_scanning.ping()`, and all fingerprinters always return the appropriate data structures, then we don't need to.

File: monkey/infection_monkey/network_scanning/ping_scanner.py [hunk 6]
```diff
@@ -6,16 +6,26 @@
...
 
 
-def ping(host: str, timeout: float) -> PingScanData:
-    if "win32" == sys.platform:
+def try_ping(host: str, timeout: float) -> PingScanData:
+    try:
+        return _ping(host, timeout)
+    except Exception:
+        logging.exception("Unhandled exception occurred while running ping")
+        return EMPTY_PING_SCAN
```
Comment (mssalvatore) [inline monkey/infection_monkey/network_scanning/ping_scanner.py:24]: I don't think the function name `try_ping()` adds any value here. Just name it `ping()` and add a docstring that defines its behavior. The caller shouldn't need to worry bout the specific mechanisms that the function is using to handle errors.

File: monkey/infection_monkey/network_scanning/tcp_scanner.py [hunk 7]
```diff
@@ -11,11 +11,20 @@
 logger = logging.getLogger(__name__)
 
 POLL_INTERVAL = 0.5
+EMPTY_PORT_SCAN = {-1: PortScanData(-1, PortStatus.CLOSED, None, None)}
 
 
-def scan_tcp_ports(
+def try_scan_tcp_ports(
```
Comment (mssalvatore) [inline monkey/infection_monkey/network_scanning/tcp_scanner.py:17]: Same comment as `tyr_ping()`

File: monkey/tests/unit_tests/infection_monkey/network_scanning/test_ping.py [hunk 8]
```diff
@@ -87,7 +87,7 @@ def set_os_windows(monkeypatch):
 @pytest.mark.usefixtures("set_os_linux")
 def test_linux_ping_success(patch_subprocess_running_ping_with_ping_output):
     patch_subprocess_running_ping_with_ping_output(LINUX_SUCCESS_OUTPUT)
-    result = ping("192.168.1.1", 1.0)
+    result = _ping("192.168.1.1", 1.0)
```
Comment (mssalvatore) [inline monkey/tests/unit_tests/infection_monkey/network_scanning/test_ping.py:90]: Unit tests should avoid calling internal functions and exercise the public interface instead.

File: monkey/tests/unit_tests/infection_monkey/master/test_ip_scanner.py [hunk 9]
```diff
@@ -202,6 +202,47 @@ def test_scan_lots_of_ips(callback, scan_config, stop):
...
+        side_effect=Exception("Exception raised when scanning TCP ports.")
+    )
+
+    ns = IPScanner(puppet, num_workers=1)
+    ns.scan(addresses, scan_config, callback, stop)
+
+    (_, scan_results) = callback.call_args_list[0][0]
+    assert scan_results.port_scan_data == {-1: PortScanData(-1, PortStatus.CLOSED, None, None)}
+
+
```
Comment (mssalvatore) [inline monkey/tests/unit_tests/infection_monkey/master/test_ip_scanner.py:232]: I don't think these two tests valid. If the ping and tcp scan operations never raise an exception, than neither will the puppet.

File: monkey/infection_monkey/network_scanning/ping_scanner.py (hunk 6 above)
Comment (VakarisZ) [reply monkey/infection_monkey/network_scanning/ping_scanner.py:24]: `try_` defines an exception wrapper. It doesn't expose anything about the mechanism, just that it's an exception handling wrapper

File: monkey/infection_monkey/master/automated_master.py [hunk 10]
```diff
@@ -225,8 +225,13 @@ def _run_plugins(
 
         interrupted_message = f"Received a stop signal, skipping remaining {plugin_type}s"
         for p in interruptible_iter(plugins, self._stop, interrupted_message):
-            # TODO: Catch exceptions to prevent thread from crashing
-            callback(p)
+            try:
+                callback(p)
+            except Exception:
+                logging.exception(
```
Comment (shreyamalviya) [inline monkey/infection_monkey/master/automated_master.py:231]:
```suggestion
                logger.exception(
```

File: monkey/infection_monkey/network_scanning/ping_scanner.py [hunk 11]
```diff
@@ -6,16 +6,26 @@
...
 logger = logging.getLogger(__name__)
 
 
-def ping(host: str, timeout: float) -> PingScanData:
-    if "win32" == sys.platform:
+def try_ping(host: str, timeout: float) -> PingScanData:
+    try:
+        return _ping(host, timeout)
+    except Exception:
+        logging.exception("Unhandled exception occurred while running ping")
```
Comment (shreyamalviya) [inline monkey/infection_monkey/network_scanning/ping_scanner.py:23]:
```suggestion
        logger.exception("Unhandled exception occurred while running ping")
```

File: monkey/infection_monkey/network_scanning/tcp_scanner.py [hunk 12]
```diff
@@ -11,11 +11,20 @@
...
     host: str, ports_to_scan: Iterable[int], timeout: float
 ) -> Mapping[int, PortScanData]:
+    try:
+        return _scan_tcp_ports(host, ports_to_scan, timeout)
+    except Exception:
+        logging.exception("Unhandled exception occurred while trying to scan tcp ports")
```
Comment (shreyamalviya) [inline monkey/infection_monkey/network_scanning/tcp_scanner.py:23]:
```suggestion
        logger.exception("Unhandled exception occurred while trying to scan tcp ports")
```

File: monkey/infection_monkey/puppet/puppet.py [hunk 13]
```diff
@@ -54,8 +56,14 @@ def fingerprint(
         port_scan_data: Dict[int, PortScanData],
         options: Dict,
     ) -> FingerprintData:
-        fingerprinter = self._plugin_registry.get_plugin(name, PluginType.FINGERPRINTER)
-        return fingerprinter.get_host_fingerprint(host, ping_scan_data, port_scan_data, options)
+        try:
+            fingerprinter = self._plugin_registry.get_plugin(name, PluginType.FINGERPRINTER)
+            return fingerprinter.get_host_fingerprint(host, ping_scan_data, port_scan_data, options)
+        except Exception:
+            logging.exception(
```
Comment (shreyamalviya) [inline monkey/infection_monkey/puppet/puppet.py:63]:
```suggestion
            logger.exception(
```

**Review by ilija-lazoroski — APPROVED** [review]
LGTM!

**Pull request merged**